arducam driver

KantaTamura opened the pull request:
[arducam driver](https://github.com/ArduCAM/Arducam-Pivariety-V4L2-Driver)をyoctoに移植した

動作確認はできていないです

KantaTamura committed:
zero/meta-arducam: Enable arducam driver

This commit add a dts overlay file and enable driver to use the
Arducam camera.

The DTS and driver for the Arducam camera have been merged into
the raspberrypi linux kernel as of version 5.15.38

Signed-off-by: KantaTamura <tkanta496@gmail.com>
Co-authored-by:: Takuya Sasaki <takuya.sasaki@spacecubics.com>
Signed-off-by: Takuya Sasaki <takuya.sasaki@spacecubics.com>

yashi commented on zero/meta-arducam/recipes-kernelmodule/arducam_0.bb:
```diff
@@ -0,0 +1,12 @@
+           file://COPYING"
+
+S = "${WORKDIR}"
+RPROVIDES_${PN} += "kernel-module-arducam"
```
nitpick: no EOF newline

yashi commented on zero/meta-arducam/recipes-kernelmodule/files/Makefile:
```diff
@@ -0,0 +1,12 @@
+	$(MAKE) -C $(KERNEL_SRC) M=$(SRC) modules_install
+
+clean:
+	$(MAKE) -C $(KERNEL_SRC) M=$(SRC) clean
```
ditto

yashi commented on zero/meta-arducam/recipes-kernelmodule/files/arducam.c:
> Note: Since [5.15.38](https://github.com/raspberrypi/firmware/commit/1441400b6b3bfeac36422a636776c60b8983a7d3), the [arducam-pivariety driver](https://github.com/raspberrypi/linux/blob/rpi-5.15.y/drivers/media/i2c/arducam-pivariety.c) has been merged into the Raspberry Pi kernel

ってあるけど、 Kirkstone の `meta-raspberrypi` だと 5.13.34 であることを commit log に記載しよう
```
$ grep LINUX_VERSION meta-raspberrypi/recipes-kernel/linux/linux-raspberrypi_5.15.bb
LINUX_VERSION ?= "5.15.34"
```

あとは FYI: https://github.com/raspberrypi/linux/pull/5033

yashi reviewed (changes requested):
良さげです。

なぜドライバーを追加する必要があるのか、だけ commit message に書いてください。コピーしてきた版の commit hash も入れておいてもらえると助かります。 

あとは  @MayaAbe か @sasataku のテスト結果が OKなら、マージで。

yashi commented:
@MayaAbe のテスト待ち

KantaTamura commented:
末尾に改行を加えた結果md5の値が変わっていたので，その部分のみ修正しました

MayaAbe commented:
カメラの動確のためにlibcameraを入れられますか？
@KantaTamura

sasataku committed:
zero/meta-arducam: Modify the kernel configuration

If we build with core-image-minimal, we will need to enable some
kernel configurations.

For configurations that should be enabled by at startup, CONFIG_XX
is set to 'y' with reference to the following page.

  - https://docs.yoctoproject.org/singleindex.html#changing-the-configuration

Kernel modules loaded from other drivers are added in layer.conf.

Signed-off-by: Takuya Sasaki <takuya.sasaki@spacecubics.com>

sasataku committed:
zero/meta-arducam: Install libarducam_pivariety.so from Arducam SDK

This commit adds pre-built libarducam_pivariety.so taken from
Arducam SDK.

Arducam has extended libcamera source code and has published it on
GitHub. However, the source code needs libarducam_pivariety.so and
it's header file to build and run.

   - repository: https://github.com/ArduCAM/arducam_ppa
   - package name: arducam-pivariety-sdk-dev
   - hash: 332543a
   - file: pool/main/a/arducam-pivariety-sdk-dev/\
           arducam-pivariety-sdk-dev_1.0.6_armhf.deb
   - file lst: usr/include/arducam/arducam_pivariety.hpp
               usr/lib/libarducam_pivariety.so.1.0.6
               usr/lib/pkgconfig/arducam_pivariety.pc

Libcamera build system utilizes pkgconfig and pkgconfig utilizes the
first .pc, we are adding the .pc file to -dev package as well.

The arducam-pivariety_1.0.6.bb was writen referencing the following
page: "Working with Pre-Built Libraries"

  https://docs.yoctoproject.org/dev/dev-manual/prebuilt-libraries.html

Signed-off-by: Takuya Sasaki <takuya.sasaki@spacecubics.com>

sasataku committed:
zero/meta-arducam: Add libcamera provided Arducam

This commit adds libcamera library provided by Arducam as
repository below.

  - https://github.com/ArduCAM/libcamera.git

In meta-openembedded, there exists the original version receipe of
libcamera, but Arducam has independently extended libcamera and
published it on gitHub.

However, do_install fails because the libcamera version targeted
by kirkstone's libcamera recipe and the arducam libcamera version
do not match.

So, we forked the recipece from commit below and modified it

  - repository: https://github.com/openembedded/meta-openembedded
  - hash: d369384
  - subject: libcamera: Simplify pipeline configuration

Signed-off-by: Takuya Sasaki <takuya.sasaki@spacecubics.com>

sasataku commented:
kirkstone (kernel 5.15.34)で、libcameraを使ってArducam Cameraが動作したので PRを更新しています。
libcamera-appsは入れていないので、動作確認は cam コマンドのみです。

- `cam -l` (使用可能なカメラのリスト表示)
- `cam -c 1 --capture=10 -F` (10 Frame分のキャプチャ&ファイル保存)

yashi commented:
Is the library based on https://github.com/ArduCAM/Arducam_dvp?

yashi commented:
I couldn't find license information for the binary file.  Do you have it?

yashi commented on zero/meta-arducam/recipes-multimedia/libcamera-arducam/libcamera-arducam_0.0.5.bb:
```diff
@@ -0,0 +1,68 @@
+SUMMARY = "Arducam libcamera framework"
```
This is more like "Arducam fork of libcamera", no?

yashi replied:
Or if you want to keep the orignal, "Arducam fork of Linux libcamera framework"

sasataku replied:
Will change to "Arducam fork of Linux libcamera framework"

yashi commented:
> Is the library based on https://github.com/ArduCAM/Arducam_dvp?

I guess not.

yashi commented on zero/meta-arducam/recipes-multimedia/libcamera-arducam/libcamera-arducam_0.0.5.bb:
```diff
@@ -0,0 +1,68 @@
+    file://LICENSES/LGPL-2.1-or-later.txt;md5=2a4f4fd2128ea2f65047ee63fbca9f68 \
+"
+
+SRC_URI = " \
```
This is the diff between this file and the original.  Is it still better to fork it instead of overlay?

```diff
diff -u meta-multimedia/recipes-multimedia/libcamera/libcamera_0.1.0.bb ~/work/launch2022/rpi-zero/zero/meta-arducam/recipes-multimedia/libcamera-arducam/libcamera-arducam_0.0.5.bb 
--- meta-multimedia/recipes-multimedia/libcamera/libcamera_0.1.0.bb	2023-09-18 21:42:16.622746910 +0900
+++ /home/yashi/work/launch2022/rpi-zero/zero/meta-arducam/recipes-multimedia/libcamera-arducam/libcamera-arducam_0.0.5.bb	2023-09-18 21:32:04.688421677 +0900
@@ -1,4 +1,4 @@
-SUMMARY = "Linux libcamera framework"
+SUMMARY = "Arducam libcamera framework"
 SECTION = "libs"
 
 LICENSE = "GPL-2.0-or-later & LGPL-2.1-or-later"
@@ -9,16 +9,16 @@
 "
 
 SRC_URI = " \
-        git://git.libcamera.org/libcamera/libcamera.git;protocol=https;branch=master \
+        git://github.com/ArduCAM/libcamera.git;protocol=https;branch=arducam \
 "
 
-SRCREV = "960d0c1e19feaf310321c906e14bd5410c6be629"
+SRCREV = "bde06144d10d4a201e761218ba9169d7a00446e6"
 
 PE = "1"
 
 S = "${WORKDIR}/git"
 
-DEPENDS = "python3-pyyaml-native python3-jinja2-native python3-ply-native python3-jinja2-native udev gnutls chrpath-native libevent libyaml"
+DEPENDS = "python3-pyyaml-native python3-jinja2-native python3-ply-native python3-jinja2-native udev gnutls chrpath-native libevent libyaml arducam-pivariety"
 DEPENDS += "${@bb.utils.contains('DISTRO_FEATURES', 'qt', 'qtbase qtbase-native', '', d)}"
 
 PACKAGES =+ "${PN}-gst"
@@ -26,10 +26,9 @@
 PACKAGECONFIG ??= ""
 PACKAGECONFIG[gst] = "-Dgstreamer=enabled,-Dgstreamer=disabled,gstreamer1.0 gstreamer1.0-plugins-base"
 
-LIBCAMERA_PIPELINES ??= "auto"
-
 EXTRA_OEMESON = " \
-    -Dpipelines=${LIBCAMERA_PIPELINES} \
+    -Dpipelines=rpi/vc4 \
+    -Dipas=rpi/vc4 \
     -Dv4l2=true \
     -Dcam=enabled \
     -Dlc-compliance=disabled \
@@ -46,8 +45,7 @@
 }
 
 do_install:append() {
-    chrpath -d ${D}${libdir}/libcamera.so
-    chrpath -d ${D}${libexecdir}/libcamera/v4l2-compat.so
+    rm -rf ${D}/usr/share
 }
 
 addtask do_recalculate_ipa_signatures_package after do_package before do_packagedata
@@ -63,10 +61,8 @@
     ${S}/src/ipa/ipa-sign-install.sh ${B}/src/ipa-priv-key.pem "${modules}"
 }
 
-FILES:${PN} += " ${libexecdir}/libcamera/v4l2-compat.so"
-FILES:${PN}-gst = "${libdir}/gstreamer-1.0"
+FILES:${PN} += " ${libdir}/v4l2-compat.so ${libdir}/libcamera/ipa*"
 
 # libcamera-v4l2 explicitly sets _FILE_OFFSET_BITS=32 to get access to
 # both 32 and 64 bit file APIs.
 GLIBC_64BIT_TIME_FLAGS = ""
-
```

sasataku replied:
Yes, I was also wondering if I should overlay with bbapend.
I will try to change to use a bbapend.

sasataku replied:
Unfortunately, do_install fails because the libcamera version targeted by kirkstone's libcamera recipe and
the arducam libcamera version  do not match.

So, we forked the recipece from commit below and modified it

yashi commented on zero/meta-arducam/recipes-multimedia/arducam-pivariety/files/arducam_pivariety.pc:
```diff
@@ -0,0 +1,19 @@
+
+Name: arducam_pivariety
+Description: Arducam pivariety camera library.
+URL: https://github.com/ArduCAM/Arducam-Pivariety-V4L2-Driver
```
I don't quite see why it points to this URL.  This `.pc` and `.hpp` are from SDK, no?

sasataku replied:
Yes, you are right.
I think it's should be points to `https://github.com/ArduCAM/arducam_ppa`.
Will update URL and description.

sasataku replied:
But, I will kept it because this is taken from https://github.com/ArduCAM/arducam_ppa.

> BTW, if the files are taken from someother places, you don't need to fix it. But please write it down where you've got them from.

yashi commented on zero/meta-arducam/recipes-multimedia/arducam-pivariety/arducam-pivariety_1.0.2.bb:
```diff
@@ -0,0 +1,34 @@
+LICENSE = "CLOSED"
+
+SRC_URI = "\
+		file://libarducam_pivariety.so \
```
Please create the symlink in `do_install()`.

sasataku replied:
OKay. will create in `do_install()`.

yashi commented on zero/meta-arducam/recipes-multimedia/arducam-pivariety/arducam-pivariety_1.0.2.bb:
```diff
@@ -0,0 +1,34 @@
+SUMMARY = "Arducam pivariety SDK recipe"
+SECTION = "libs"
+LICENSE = "CLOSED"
```
If the library is non-distributable, we have to do something.

sasataku replied:
Now I'm asking the license information in Arducam forum.
https://forum.arducam.com/t/what-is-the-license-for-arducam-pivariety-sdk-dev/5760

yashi commented on zero/meta-arducam/recipes-multimedia/arducam-pivariety/arducam-pivariety_1.0.2.bb:
```diff
@@ -0,0 +1,34 @@
+INHIBIT_PACKAGE_DEBUG_SPLIT = "1"
+
+RDEPENDS_${PN} += "libgcc zlib"
+DEPENDS += " \
+          zlib \
+          libgcc \
+"
```
Why do we need this?

sasataku replied:
If we do not explicitly specify dependencies when packaging pre-built binaries, we  will receive errors like the following:

```
ERROR: arducam-pivariety-1.0.2-r0 do_package_qa: QA Issue: /usr/lib/libarducam_pivariety.so.1.0.2 contained in package arducam-pivariety requires libz.so.1()(64bit), but no providers found in RDEPENDS:arducam-pivariety? [file-rdeps]
```

So added the dependencies.

yashi commented on zero/meta-arducam/recipes-multimedia/arducam-pivariety/arducam-pivariety_1.0.2.bb:
```diff
@@ -0,0 +1,34 @@
+}
+
+FILES:${PN} = "${libdir}/libarducam_pivariety.so**"
+FILES:${PN}-dev = "${libdir}/pkgconfig/arducam_pivariety.pc \
+                    ${includedir}/arducam"
```
Missed `.hpp` in `-dev`?

sasataku replied:
Thanks. will fix.

yashi commented on zero/meta-arducam/recipes-multimedia/arducam-pivariety/arducam-pivariety_1.0.2.bb:
```diff
@@ -0,0 +1,34 @@
+		file://arducam_pivariety.pc \
+"
+
+INHIBIT_PACKAGE_STRIP = "1"
+INHIBIT_SYSROOT_STRIP = "1"
+INHIBIT_PACKAGE_DEBUG_SPLIT = "1"
```
care to share why?

sasataku replied:
The pre-build binary is already stripped the debug information.
So if we don't set it, receive the error below.

> INHIBIT_PACKAGE_STRIP = "1"

```
ERROR: arducam-pivariety-1.0.2-r0 do_package: QA Issue: File '/usr/lib/libarducam_pivariety.so.1.0.2' from arducam-pivariety was already stripped, this will prevent future debugging! [already-stripped]
```

> INHIBIT_SYSROOT_STRIP

```
WARNING: arducam-pivariety-1.0.2-r0 do_populate_sysroot: File '/usr/lib/libarducam_pivariety.so.1.0.2' from arducam-pivariety was already stripped, this will prevent future debugging!
ERROR: arducam-pivariety-1.0.2-r0 do_package: QA Issue: File '/usr/lib/libarducam_pivariety.so.1.0.2' from arducam-pivariety was already stripped, this will prevent future debugging! [already-stripped]
```

> INHIBIT_PACKAGE_DEBUG_SPLIT

```
WARNING: arducam-pivariety-1.0.2-r0 do_populate_sysroot: File '/usr/lib/libarducam_pivariety.so.1.0.2' from arducam-pivariety was already stripped, this will prevent future debugging!
ERROR: arducam-pivariety-1.0.2-r0 do_package: QA Issue: File '/usr/lib/libarducam_pivariety.so.1.0.2' from arducam-pivariety was already stripped, this will prevent future debugging! [already-stripped]
```

yashi commented on zero/meta-arducam/conf/layer.conf:
```diff
@@ -6,6 +6,7 @@ BBFILES += "\
     ${LAYERDIR}/recipes-devicetree/linux-raspberrypi_%.bbappend \
```
I don't know where to put my comment on the commit log so here goes.
- please check english grammer.
- The arducam_ppa provides packages and there is a package providing a pre-built "binary".
- The word "various" isn't the right word because you are installing `.so` and `.hpp`.  You should generate the symlink.
- this "Libcamera's build system utilizes the first pkgconfig to search for this SDK, so it also installs .pc file." isn't right.  libcamera uses pkgconfig (or pkg-config).  pkgconfig uses the first matched `.pc` file (if it's true; I haven't tested).
- This "It used the following page as a reference for how to install pre-built packages with Yocto." is unbigous.  Since we were just talking about "Libcamera's build system", "pkgconfig", or the `.pc` file, the word "it" sounds like referencing one of them.  Instead I'm sure you are referencing the `.bb` file.

yashi replied:
my take:

> zero/meta-arducam: Install libarducam_pivariety.so.1.0.2 from Arducam SDK
> 
> This commit adds pre-built libarducam_pivariety.so taken from Arducam SDK.
> 
> Arducam has extended libcamera source code and has published it on
> GitHub. However, the source code needs libarducam_pivariety.so and
> it's header file to build and run.
> 
>   - https://github.com/ArduCAM/arducam_ppa
>     arducam-pivariety-sdk-dev
> 
> (We should add SHA hash or tags where we obtained files)
> (We should add list of files taken from the package)
> 
> Libcamera build system utilizes pkgconfig and pkgconfig utilizes the
> first .pc, we are adding the .pc file to -dev package as well.
> 
> The arducam-pivariety_1.0.2.bb was writen referencing the following
> page: "Working with Pre-Built Libraries"
> 
>   - https://docs.yoctoproject.org/dev/dev-manual/prebuilt-libraries.html

Please update hash value, tags, package name, list of files taken.

sasataku replied:
Thanks. will update commit log and add SHA hash.etc.

yashi commented:
BTW, if the files are taken from someother places, you don't need to fix it.  But please write it down where you've got them from.
